Make contributing easier

Estus-Dev opened the pull request:
I've added a [Contributing section to the README](./README.md#Contributing), detailing how to add new programs to the list without being rejected by CI.

I've also fixed a bug that's bitten me both times I've contributed in the past, where `scripts/update-chip8-archive.js` would inject invalid colors into existing entries, causing schema validation to fail.

This was caused by the Chip8Archive's `programs.json` file containing programs where no colors were specified.

In future I'd like to automatically create _(draft)_ PRs for updates to the Chip8Archive, at which point we could remove that source of friction from new manual PRs and simplify the README I just added.

Estus-Dev committed:
fix: Don't generate invalid colors when absent

Previously when parsing data from the chip8Archive,  this was creating
an object that looked like the following on mini-lights-out and tanks,
which have no colors specified in the chip8Archive JSON data:

```json
colors: {
    pixels: ["missing color", "missing color"]
}
```

That object caused schema validation to fail and needed to be cleaned
up manually before making a PR.

I'm fixing this so that the contributing guide I'm writing doesn't have
to include a manual step in order to pass CI.

I've also tried to match the existing style of the script, with no blank
lines inside a function.

Estus-Dev committed:
Add Contributing section to the end of the README

Estus-Dev committed:
Pure vanity

Estus-Dev committed:
fix: Missing link from footer cleanup

I was using footer links out of habit when I started this PR, but
switched to inline links to match the rest of the document. Looks like I
missed this one in the switch.

Timendus commented on README.md:
```diff
@@ -231,3 +231,55 @@ are some places you can find the games listed in this database:
+Once NPM is installed, navigate to the root of this repository and execute
+`npm install` to install this project's dependencies.
+
+### Updating the Database from Octo
+
+The CI system checks to ensure that the database is up to date with the contents
+of the [CHIP-8 Archive], and will fail if there are programs there that are not
+listed in [`programs.json`](./database/programs.json).
+
+Before making your changes, it is advised that you run `npm start` to
+automatically create entries for the new files.
+
+If there are any changes (check with `git status` or `git diff`), review the
+newly generated entries at the end of
+[`programs.json`](./database/programs.json) and perform any cleanup that needs
+to occur. In particular, determine appropriate values for the `"platforms"`
+section for each newly added program.
+
+If you made any changes, run `npm run format` before committing.
```
I don't think this is necessary. The Github CI "only" checks to see if everything is formatted correctly, and passes the schema. It does this:

https://github.com/chip-8/chip-8-database/blob/bff63251ff9d0c95b2aab4c4408f92fb4a8d23f9/.github/workflows/test.yaml#L29-L31

Which do:

https://github.com/chip-8/chip-8-database/blob/bff63251ff9d0c95b2aab4c4408f92fb4a8d23f9/package.json#L7-L8

So that's just a Prettier check, and invoking the `validate.js` script, which does this, according to the comment:

https://github.com/chip-8/chip-8-database/blob/bff63251ff9d0c95b2aab4c4408f92fb4a8d23f9/scripts/validate.js#L1-L3

I wouldn't want to burden new contributors with having to update the CHIP-8 Archive before they can add their own programs.

Edit: which should also lessen the issue of the colours getting messed up by the import script. That's a script to be ran by maintainers, not random contributors, IMO. But still great to fix the issue.

Estus-Dev replied:
I also don't like that being there and would be glad to remove it. In fact I was planning on adding a weekly CI run to create a PR a la dependabot and removing it then. That PR would need to be a draft for manual review of platforms/etc... but it'd be a nice prompt to action.

The reason I put it there was because I had that issue both times when contributing: because I ran `npm start` to generate hashes. That's the default command I try when I know there are `package.json` scripts to be run.

What do you think about about moving `scripts.start` to `scripts.update`, and not having an `npm start`?

Estus-Dev replied:
Or alternatively having `npm start` run `npm run generate-hashes-file && npm run format`?

Timendus replied:
That sounds like a great idea 👍

Timendus commented on README.md:
```diff
@@ -231,3 +231,55 @@ are some places you can find the games listed in this database:
+[`schemas/programs.json`](./schemas/programs.json) for type definitions and
+validation info.
+
+Otherwise you can look at existing entries and use `npm run test` to
+automatically validate against the schema.
+
+Once you're done making your changes, run `npm start` and `npm run format`
+before committing.
```
Maybe just say:

> Once you're done making your changes, run `npm run format` and make sure `npm test` passes before committing.

Estus-Dev replied:
IMO it also would help new contributors if `npm test` is all you need to ensure you pass CI. Could we rename `scripts.test` to `scripts.validate-schema` to free it up for `npm run validate-schema && npm run check-formatting`?

Estus-Dev replied:
> run `npm run format` and make sure `npm test` passes before committing.

As an aside: If I shorten it to those two commands, then the SHA-1 file won't be updated.

Timendus replied:
Ah, yes. It also needs a `generate-hashes-file` in there. Good catch. I'm fine with moving a couple of commands around if that makes things easier for newcomers 💪

Timendus reviewed (changes requested):
Looking great. One big remark below.

Estus-Dev committed:
Specify why you should add new entries to the end

Estus-Dev committed:
Link to CHIP-8 Archive

Estus-Dev committed:
Change package.json scripts to simplify the README

Estus-Dev committed:
Keep Update section but move it to the end

Estus-Dev commented:
EDIT: Didn't refresh the page before commenting.

I figured it'd be easier to make the changes I proposed than to get signoff first and then make them and then have you review them again.

In particular, I moved the updating section to the bottom of the page because it's still useful to have documented, but I removed the incorrect info about CI.

I changed the commands around as I suggested, so `npm start` and `npm test` are all a new contributor needs to know, while leaving `npm run update` for pulling from upstream.
